cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/MacroParser.java

```diff
@@ -0,0 +1,270 @@
+   *
+   * @param str the string to find a macro in
+   * @return the rightmost macro and its position in the string
+   * @throws InvalidMacroException if invalid macro syntax was found. This cannot be thrown if isLenient is true.
+   */
+  @Nullable
+  private static MacroMetadata findRightmostMacro(String str, @Nullable MacroEvaluator macroEvaluator,
+                                                  @Nullable Set<String> macros) {
+    // find opening "${" skipping escaped syntax "\${" TODO: and allowing doubly-escaping "\\${"
+    int startIndex = getStartIndex(str);
+    if (startIndex < 0) {
+      return null;
+    }
+
+    // found "${", now look for enclosing "}" and allow escaping "\}" TODO: and doubly-escaping "\\}"
+    int endIndex = getEndIndex(str, startIndex);
+    if (endIndex < 0) {
+      throw new InvalidMacroException(String.format("Could not find enclosing '}' for macro '%s'.",
+                                                    str.substring(startIndex)));
+    }
+
+    // macroStr = 'macroFunction(macroArguments)' or just 'property' for ${property}
+    String macroStr = str.substring(startIndex + 2, endIndex).trim();
+
+    // look for "(", which indicates there are arguments and allow escaping "\(" TODO: and doubly-escaping "\\("
+    int argsStartIndex = getArgsStartIndex(str, macroStr);
+
+    // determine whether to use a macro function or a property lookup
+    if (argsStartIndex >= 0) {
+      return getMacroFunctionMetadata(macroEvaluator, macros, startIndex, endIndex, macroStr, argsStartIndex);
+    } else {
+      return getMacroLookupMetadata(macroEvaluator, macros, startIndex, endIndex, macroStr);
+    }
+  }
+
+  private static MacroMetadata getMacroFunctionMetadata(@Nullable MacroEvaluator macroEvaluator,
+                                                        @Nullable Set<String> macros, int startIndex, int endIndex,
+                                                        String macroStr, int argsStartIndex) {
+    // check for closing ")" and allow escaping "\)" TODO: and doubly-escaping "\\)"
+    int closingParenIndex = getClosingParenIndex(macroStr);
+    if (closingParenIndex < 0 || !macroStr.endsWith(")")) {
+      throw new InvalidMacroException(String.format("Could not find enclosing ')' for macro arguments in '%s'.",
+                                                    macroStr));
+    }
+
+    // arguments and macroFunction are expected to have escapes replaced when being evaluated
+    String arguments = macroStr.substring(argsStartIndex + 1, macroStr.length() - 1);
+    String macroFunction = macroStr.substring(0, argsStartIndex);
+    if (macroEvaluator != null) {
+      arguments = replaceEscapedSyntax(arguments);
+      macroFunction = replaceEscapedSyntax(macroFunction);
+    }
+    String[] args = COMMA.split(arguments);
+    if (macros != null) {
+      for (String argument : args) {
+        macros.add(argument);
```

Comment:
ok i see these aren't passed to the actual function, but only used for the getMacros() call?